File: Editorial/114_Flatten_Binary_Tree_to_Linked_List/JavaScript/3_O_Iterative_Solution.js
var flatten = function (root) {
    // Handle the null scenario
    if (!root) {
        return;
    }
    let node = root;
    while (node != null) {
        // If the node has a left child
        if (node.left != null) {
            // Find the rightmost node
            let rightmost = node.left;
            while (rightmost.right != null) {
                rightmost = rightmost.right;
            }
            // rewire the connections
            rightmost.right = node.right;
            node.right = node.left;
            node.left = null;
        }
        // move on to the right side of the tree
        node = node.right;
    }
};

File: Editorial/114_Flatten_Binary_Tree_to_Linked_List/Go/3_O_Iterative_Solution.go
func flatten(root *TreeNode) {
    // Handle the null scenario
    if root == nil {
        return
    }
    node := root
    for node != nil {
        // If the node has a left child
        if node.Left != nil {
            // Find the rightmost node
            rightmost := node.Left
            for rightmost.Right != nil {
                rightmost = rightmost.Right
            }
            // rewire the connections
            rightmost.Right = node.Right
            node.Right = node.Left
            node.Left = nil
        }
        // move on to the right side of the tree
        node = node.Right
    }
}
